addons/website/controllers/model_page.py

```diff
@@ -0,0 +1,134 @@
+import ast
+
+import werkzeug
+
+from odoo.addons.http_routing.models.ir_http import slug, unslug
+from odoo.http import Controller, request, route
+from odoo.osv.expression import AND, OR
+
+
+class ModelPageController(Controller):
+    pager_step = 20
+
+    @route([
+        "/model/<string:page_name_slugified>",
+        "/model/<string:page_name_slugified>/page/<int:page_number>",
+        "/model/<string:page_name_slugified>/<string:record_slug>",
+    ], website=True, auth="public")
+    def generic_model(self, page_name_slugified=None, page_number=1, record_slug=None, **searches):
+        if not page_name_slugified:
+            raise werkzeug.exceptions.NotFound()
+
+        website = request.website
+
+        page_type = "listing"
+        if record_slug is not None:
+            page_type = "single"
+
+        website_page_domain = AND([
+            [("page_type", "=", page_type)],
+            [("name_slugified", "=", page_name_slugified)],
+            [("website_published", "=", True)],
+            website.website_domain(),
+        ])
+
+        page = request.env["website.controller.page"].sudo().search(website_page_domain, limit=1)
+        if not page:
+            raise werkzeug.exceptions.NotFound()
+
+        view = page.view_id
+        if not view:
+            raise werkzeug.exceptions.NotFound()
+
+        target_model_name = page.model_id.model
+        Model = request.env[target_model_name]
+        if not Model.check_access_rights("read", raise_exception=False):
+            raise werkzeug.exceptions.NotFound()
+
+        rec_domain = ast.literal_eval(page.record_domain or "[]")
+        domains = [rec_domain]
+
+        if record_slug:
+            _, res_id = unslug(record_slug)
+            record = Model.browse(res_id).filtered_domain(AND(domains))
+            # We check for slug matching because we are not entirely sure
+            # that we end up seeing record for the right model
+            # i.e. in case of a redirect when a "single" page doesn't match the listing
+            if not record.exists() or record_slug != slug(record):
+                raise werkzeug.exceptions.NotFound()
+
+            listing = request.env["website.controller.page"].sudo().search(AND([
+                    [("name_slugified", "=", page_name_slugified)],
+                    [("page_type", "=", "listing")],
+                    [("model", "=", target_model_name)],
+                    [("website_published", "=", True)],
+                    website.website_domain(),
+                ]))
```

Comment:
Wrong indent (and same for website_published as above)
![image](https://github.com/odoo/odoo/assets/30048408/8df99805-12aa-4196-8f63-2e1dfe81b8d2)